codalab/model/bundle_model.py

```diff
@@ -2901,6 +2901,25 @@ def find_oauth2_token(self, client_id, user_id, expires_after):
 
         return OAuth2Token(self, **row)
 
+    def access_token_exists_for_user(self, client_id: str, user_id: str, access_token: str) -> bool:
```

Comment:
rename to `check_access_token`?

Reply:
maybe just replace with `find_oauth2_token(...).access_token == access_token`

Reply:
Done.

Reply:
OH, wait! I remember why I needed to do this. Basically, `find_oauth2_token` just fetches a single `oauth2` token, but the user may have several valid tokens active. For instance, here's my local `oauth2_token` database for the codalab user:
```
mysql> select * from oauth2_token where user_id=0
    -> ;
+-----+-----------------------+---------+---------+--------------------------------+--------------------------------+---------------------+
| id  | client_id             | user_id | scopes  | access_token                   | refresh_token                  | expires             |
+-----+-----------------------+---------+---------+--------------------------------+--------------------------------+---------------------+
| 123 | codalab_cli_client    | 0       | default | tDCP52XbDLP37nujTJOOjNzTS6Y68r | SVaPGWY9PpIA4JcLaHYfK0dvvJ48hD | 2023-08-16 06:37:25 |
| 124 | codalab_worker_client | 0       | default | F3tES3W9SEvpRfzW1d07CpzCzydYpe | 9fBYlZNpwmnH9TQTVmprH3YYae89v2 | 2023-08-16 06:37:41 |
| 125 | codalab_cli_client    | 0       | default | thgAVTJSNokta4g7EwZvWQNsmGDhZV | jmGC6hjBnB08EdujJ3qeH2Pt3Y5QOM | 2023-08-16 06:44:27 |
| 126 | codalab_cli_client    | 0       | default | ReKNdvycOdWjhae0PFubFRk8FAADxr | y5HiM5nIlxGpM4eSgZWJmOpX7RC1Lv | 2023-08-16 06:46:38 |
| 127 | codalab_worker_client | 0       | default | DNIvpmrPLmqK6o3ZEgYLYFF0HKkygO | VlOnzl1HfppROOhXQcDIKtW0hIy5rO | 2023-08-16 06:46:52 |
| 128 | codalab_cli_client    | 0       | default | UOYOfCkaU8y4dX8y3gczRaQmzbHTnQ | 0i9caPdfe5tJ2SU21tWIuZG9uFVfkO | 2023-08-16 06:47:11 |
| 129 | codalab_cli_client    | 0       | default | o36T83O7raUDwmDqywJijl2sLwwq5a | RL5MJ8W25UEgPiA23IWL8U0FWmc8Ob | 2023-08-16 06:49:26 |
| 130 | codalab_worker_client | 0       | default | qEiksslyxtsfuAxHJraqFR1gnjXhYT | MZug0gzJHv5uMwP1MhaSOlPRZdXjXg | 2023-08-16 06:49:59 |
| 131 | codalab_cli_client    | 0       | default | 5osMISohDoyVhDIOkjMO4VStyODgBU | Eb11XOj0pGhV6PzhKSZOKJSp2xFqfM | 2023-08-16 06:51:35 |
| 132 | codalab_worker_client | 0       | default | 9cttmuOkfXXUwawcmqXKu1g3Q3jOWJ | OO2Ey0ukIIN2Rm3zKUGLB5M4PMmbXy | 2023-08-16 06:51:59 |
| 133 | codalab_cli_client    | 0       | default | lDbuVCNK3Oee9Zlvv2n32aXjm5yI5V | dMJR0evB2fXUmkZgSotDLZxuKTBvqV | 2023-08-16 06:54:22 |
| 134 | codalab_worker_client | 0       | default | 8dq76r4bph03KKT4S06Gx9sMxAJyfL | OZChA69SQ470M84LH18w8fNg0sIkPi | 2023-08-16 06:54:46 |
| 135 | codalab_cli_client    | 0       | default | 43RJxmkngMqehMv4JyqMZUqif7ZSxi | xnerdfwqlyYEkuXiAm89FTPvoRjzjL | 2023-08-16 06:59:25 |
| 136 | codalab_worker_client | 0       | default | PDmOLXvJif5Rj6jdky3oC46CpMOnCi | BMQzvyzVbloRYauj2eVC7Xo7mC4YYE | 2023-08-16 06:59:38 |
+-----+-----------------------+---------+---------+--------------------------------+--------------------------------+---------------------+
14 rows in set (0.00 sec)
```

When I use `find_oauth2_token`, it returns the Oauth2Token with access token `F3tES3W9SEvpRfzW1d07CpzCzydYpe`, but the worker sends the access token `PDmOLXvJif5Rj6jdky3oC46CpMOnCi`. Both are valid, the auth is rejected when I use `find_oauth2_token`.

As such, we have two options: keep the function I made or modify `find_oauth2_token` to return all the valid tokens and then search through that list for the token we want.

For now, I'll leave in my earlier change and do that.

Reply:
I'm also not sure why the user has so many tokens active -- perhaps I should dig deeper into the Oauth2 code.